File: backend/services/enterprise_service.go
package services

import (
	"ai-project-backend/database"
	"ai-project-backend/models"
	"context"
	"crypto/rand"
	"fmt"
	"math/big"
)

// EnterpriseService provides business logic for managing enterprises
type EnterpriseService struct {
	enterpriseRepo database.EnterpriseRepository
	auditLogger    *AsyncAuditLogger
}

// NewEnterpriseService creates a new enterprise service
func NewEnterpriseService(enterpriseRepo database.EnterpriseRepository, auditLogger *AsyncAuditLogger) *EnterpriseService {
	return &EnterpriseService{
		enterpriseRepo: enterpriseRepo,
		auditLogger:    auditLogger,
	}
}

// Enterprise operations

// CreateEnterprise creates a new enterprise
func (s *EnterpriseService) CreateEnterprise(ctx context.Context, req *models.EnterpriseRequest, operatorID int) (*models.Enterprise, error) {
	// Check if code already exists
	existing, err := s.enterpriseRepo.GetByCode(ctx, req.Code)
	if err == nil && existing != nil {
		return nil, fmt.Errorf("enterprise code '%s' already exists", req.Code)
	}

	// Create enterprise object
	enterprise := &models.Enterprise{
		Name:                req.Name,
		Code:                req.Code,
		Description:         req.Description,
		IndustryType:        req.IndustryType,
		BusinessType:        req.BusinessType,
		RegistrationNumber:  req.RegistrationNumber,
		TaxID:              req.TaxID,
		LegalRepresentative: req.LegalRepresentative,
		ContactEmail:        req.ContactEmail,
		ContactPhone:        req.ContactPhone,
		Address:             req.Address,
		City:                req.City,
		Province:            req.Province,
		PostalCode:          req.PostalCode,
		Website:             req.Website,
		Status:              req.Status,
		CreatedBy:           &operatorID,
	}

	// Create enterprise
	createdEnterprise, err := s.enterpriseRepo.Create(ctx, enterprise)
	if err != nil {
		return nil, fmt.Errorf("failed to create enterprise: %w", err)
	}

	// Log the creation event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "create",
		ResourceType: "enterprise",
		ResourceID:   fmt.Sprintf("%d", createdEnterprise.ID),
		Description:  fmt.Sprintf("Created enterprise %s (%s)", req.Name, req.Code),
		Status:       "success",
	})

	return createdEnterprise, nil
}

// GetEnterpriseByID gets an enterprise by ID
func (s *EnterpriseService) GetEnterpriseByID(ctx context.Context, id int) (*models.Enterprise, error) {
	enterprise, err := s.enterpriseRepo.GetByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("enterprise not found: %w", err)
	}

	return enterprise, nil
}

// GetEnterpriseByCode gets an enterprise by code
func (s *EnterpriseService) GetEnterpriseByCode(ctx context.Context, code string) (*models.Enterprise, error) {
	enterprise, err := s.enterpriseRepo.GetByCode(ctx, code)
	if err != nil {
		return nil, fmt.Errorf("enterprise not found: %w", err)
	}

	return enterprise, nil
}

// UpdateEnterprise updates an enterprise
func (s *EnterpriseService) UpdateEnterprise(ctx context.Context, id int, req *models.EnterpriseUpdateRequest, operatorID int) (*models.Enterprise, error) {
	// Get existing enterprise
	enterprise, err := s.enterpriseRepo.GetByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("enterprise not found: %w", err)
	}

	// Check if code is being changed and conflicts with existing
	if req.Code != nil && *req.Code != enterprise.Code {
		existing, err := s.enterpriseRepo.GetByCode(ctx, *req.Code)
		if err == nil && existing != nil && existing.ID != id {
			return nil, fmt.Errorf("enterprise code '%s' already exists", *req.Code)
		}
	}

	// Update fields only if provided
	if req.Name != nil {
		enterprise.Name = *req.Name
	}
	if req.Code != nil {
		enterprise.Code = *req.Code
	}
	if req.Description != nil {
		enterprise.Description = req.Description
	}
	if req.IndustryType != nil {
		enterprise.IndustryType = req.IndustryType
	}
	if req.BusinessType != nil {
		enterprise.BusinessType = *req.BusinessType
	}
	if req.RegistrationNumber != nil {
		enterprise.RegistrationNumber = req.RegistrationNumber
	}
	if req.TaxID != nil {
		enterprise.TaxID = req.TaxID
	}
	if req.LegalRepresentative != nil {
		enterprise.LegalRepresentative = req.LegalRepresentative
	}
	if req.ContactEmail != nil {
		enterprise.ContactEmail = req.ContactEmail
	}
	if req.ContactPhone != nil {
		enterprise.ContactPhone = req.ContactPhone
	}
	if req.Address != nil {
		enterprise.Address = req.Address
	}
	if req.City != nil {
		enterprise.City = req.City
	}
	if req.Province != nil {
		enterprise.Province = req.Province
	}
	if req.PostalCode != nil {
		enterprise.PostalCode = req.PostalCode
	}
	if req.Website != nil {
		enterprise.Website = req.Website
	}
	if req.Status != nil {
		enterprise.Status = *req.Status
	}
	enterprise.UpdatedBy = &operatorID

	// Update enterprise
	updatedEnterprise, err := s.enterpriseRepo.Update(ctx, enterprise)
	if err != nil {
		return nil, fmt.Errorf("failed to update enterprise: %w", err)
	}

	// Log the update event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "update",
		ResourceType: "enterprise",
		ResourceID:   fmt.Sprintf("%d", id),
		Description:  fmt.Sprintf("Updated enterprise %s (%s)", req.Name, req.Code),
		Status:       "success",
	})

	return updatedEnterprise, nil
}

// DeleteEnterprise soft deletes an enterprise
func (s *EnterpriseService) DeleteEnterprise(ctx context.Context, id int, operatorID int) error {
	// Check if enterprise exists
	enterprise, err := s.enterpriseRepo.GetByID(ctx, id)
	if err != nil {
		return fmt.Errorf("enterprise not found: %w", err)
	}

	// Delete enterprise
	err = s.enterpriseRepo.Delete(ctx, id)
	if err != nil {
		return fmt.Errorf("failed to delete enterprise: %w", err)
	}

	// Log the deletion event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "delete",
		ResourceType: "enterprise",
		ResourceID:   fmt.Sprintf("%d", id),
		Description:  fmt.Sprintf("Deleted enterprise %s (%s)", enterprise.Name, enterprise.Code),
		Status:       "success",
	})

	return nil
}

// ListEnterprises lists enterprises with pagination and filtering
func (s *EnterpriseService) ListEnterprises(ctx context.Context, limit, offset int, filters map[string]interface{}) ([]*models.Enterprise, int, error) {
	enterprises, total, err := s.enterpriseRepo.List(ctx, limit, offset, filters)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to list enterprises: %w", err)
	}

	return enterprises, total, nil
}

// GetEnterpriseStats returns statistics about enterprises
func (s *EnterpriseService) GetEnterpriseStats(ctx context.Context) (*models.EnterpriseStats, error) {
	stats, err := s.enterpriseRepo.GetStats(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get enterprise statistics: %w", err)
	}

	return stats, nil
}

// Enterprise User operations

// CreateEnterpriseUser creates a new enterprise user with generated password
func (s *EnterpriseService) CreateEnterpriseUser(ctx context.Context, req *models.EnterpriseUserRequest, operatorID int) (*models.EnterpriseUser, string, error) {
	// Validate enterprise exists
	_, err := s.enterpriseRepo.GetByID(ctx, req.EnterpriseID)
	if err != nil {
		return nil, "", fmt.Errorf("enterprise not found: %w", err)
	}

	// Check if username already exists in this enterprise
	users, _, err := s.enterpriseRepo.GetUsers(ctx, req.EnterpriseID, 1000, 0) // Get all users
	if err != nil {
		return nil, "", fmt.Errorf("failed to check existing users: %w", err)
	}

	for _, user := range users {
		if user.Username == req.Username {
			return nil, "", fmt.Errorf("username '%s' already exists in this enterprise", req.Username)
		}
	}

	// If this is set as primary contact, ensure no other primary contact exists for this enterprise
	if req.IsPrimaryContact {
		existing, err := s.GetEnterprisePrimaryContact(ctx, req.EnterpriseID)
		if err == nil && existing != nil {
			return nil, "", fmt.Errorf("enterprise already has a primary contact: %s", existing.Name)
		}
	}

	// Generate initial password
	initialPassword, err := s.generateRandomPassword()
	if err != nil {
		return nil, "", fmt.Errorf("failed to generate password: %w", err)
	}

	// Create user object
	user := &models.EnterpriseUser{
		EnterpriseID:     req.EnterpriseID,
		Username:         req.Username,
		Email:            req.Email,
		Name:             req.Name,
		Phone:            req.Phone,
		Position:         req.Position,
		IsPrimaryContact: req.IsPrimaryContact,
		AccessLevel:      req.AccessLevel,
		Status:           req.Status,
		Bio:              req.Bio,
		CreatedBy:        &operatorID,
	}

	// Create user
	createdUser, err := s.enterpriseRepo.CreateUser(ctx, user)
	if err != nil {
		return nil, "", fmt.Errorf("failed to create enterprise user: %w", err)
	}

	// Log the creation event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "create",
		ResourceType: "enterprise_user",
		ResourceID:   fmt.Sprintf("%d", createdUser.ID),
		Description:  fmt.Sprintf("Created enterprise user %s for enterprise %d", req.Username, req.EnterpriseID),
		Status:       "success",
	})

	return createdUser, initialPassword, nil
}

// GetEnterpriseUserByID gets an enterprise user by ID
func (s *EnterpriseService) GetEnterpriseUserByID(ctx context.Context, userID int) (*models.EnterpriseUser, error) {
	user, err := s.enterpriseRepo.GetUserByID(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("enterprise user not found: %w", err)
	}

	return user, nil
}

// UpdateEnterpriseUser updates an enterprise user
func (s *EnterpriseService) UpdateEnterpriseUser(ctx context.Context, userID int, req *models.EnterpriseUserRequest, operatorID int) (*models.EnterpriseUser, error) {
	// Get existing user
	user, err := s.enterpriseRepo.GetUserByID(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("enterprise user not found: %w", err)
	}

	// If setting as primary contact, ensure no other primary contact exists for this enterprise
	if req.IsPrimaryContact && !user.IsPrimaryContact {
		existing, err := s.GetEnterprisePrimaryContact(ctx, user.EnterpriseID)
		if err == nil && existing != nil && existing.ID != userID {
			return nil, fmt.Errorf("enterprise already has a primary contact: %s", existing.Name)
		}
	}

	// Update fields
	user.Username = req.Username
	user.Email = req.Email
	user.Name = req.Name
	user.Phone = req.Phone
	user.Position = req.Position
	user.IsPrimaryContact = req.IsPrimaryContact
	user.AccessLevel = req.AccessLevel
	user.Status = req.Status
	user.Bio = req.Bio
	user.UpdatedBy = &operatorID

	// Update user
	updatedUser, err := s.enterpriseRepo.UpdateUser(ctx, user)
	if err != nil {
		return nil, fmt.Errorf("failed to update enterprise user: %w", err)
	}

	// Log the update event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "update",
		ResourceType: "enterprise_user",
		ResourceID:   fmt.Sprintf("%d", userID),
		Description:  fmt.Sprintf("Updated enterprise user %s", user.Username),
		Status:       "success",
	})

	return updatedUser, nil
}

// DeleteEnterpriseUser soft deletes an enterprise user
func (s *EnterpriseService) DeleteEnterpriseUser(ctx context.Context, userID int, operatorID int) error {
	// Get user for logging
	user, err := s.enterpriseRepo.GetUserByID(ctx, userID)
	if err != nil {
		return fmt.Errorf("enterprise user not found: %w", err)
	}

	// Delete user
	err = s.enterpriseRepo.DeleteUser(ctx, userID)
	if err != nil {
		return fmt.Errorf("failed to delete enterprise user: %w", err)
	}

	// Log the deletion event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "delete",
		ResourceType: "enterprise_user",
		ResourceID:   fmt.Sprintf("%d", userID),
		Description:  fmt.Sprintf("Deleted enterprise user %s", user.Username),
		Status:       "success",
	})

	return nil
}

// GetEnterpriseUsers lists users for an enterprise
func (s *EnterpriseService) GetEnterpriseUsers(ctx context.Context, enterpriseID int, limit, offset int) ([]*models.EnterpriseUser, int, error) {
	// Validate enterprise exists
	_, err := s.enterpriseRepo.GetByID(ctx, enterpriseID)
	if err != nil {
		return nil, 0, fmt.Errorf("enterprise not found: %w", err)
	}

	users, total, err := s.enterpriseRepo.GetUsers(ctx, enterpriseID, limit, offset)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get enterprise users: %w", err)
	}

	return users, total, nil
}

// GetEnterprisePrimaryContact gets the primary contact for an enterprise
func (s *EnterpriseService) GetEnterprisePrimaryContact(ctx context.Context, enterpriseID int) (*models.EnterpriseUser, error) {
	user, err := s.enterpriseRepo.GetPrimaryContact(ctx, enterpriseID)
	if err != nil {
		return nil, fmt.Errorf("failed to get primary contact for enterprise %d: %w", enterpriseID, err)
	}

	return user, nil
}

// Enterprise Department operations

// CreateEnterpriseDepartment creates a new enterprise department
func (s *EnterpriseService) CreateEnterpriseDepartment(ctx context.Context, req *models.EnterpriseDepartmentRequest, operatorID int) (*models.EnterpriseDepartment, error) {
	// Validate enterprise exists
	_, err := s.enterpriseRepo.GetByID(ctx, req.EnterpriseID)
	if err != nil {
		return nil, fmt.Errorf("enterprise not found: %w", err)
	}

	// Validate parent department if specified
	if req.ParentID != nil {
		parent, err := s.enterpriseRepo.GetDepartmentByID(ctx, *req.ParentID)
		if err != nil {
			return nil, fmt.Errorf("parent department not found: %w", err)
		}
		if parent.EnterpriseID != req.EnterpriseID {
			return nil, fmt.Errorf("parent department belongs to different enterprise")
		}
	}

	// Create department object
	dept := &models.EnterpriseDepartment{
		EnterpriseID:  req.EnterpriseID,
		Name:          req.Name,
		ParentID:      req.ParentID,
		SortOrder:     req.SortOrder,
		ManagerID:     req.ManagerID,
		Description:   req.Description,
		EmployeeCount: req.EmployeeCount,
		Status:        req.Status,
		CreatedBy:     &operatorID,
	}

	// Create department
	createdDept, err := s.enterpriseRepo.CreateDepartment(ctx, dept)
	if err != nil {
		return nil, fmt.Errorf("failed to create enterprise department: %w", err)
	}

	// Log the creation event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "create",
		ResourceType: "enterprise_department",
		ResourceID:   fmt.Sprintf("%d", createdDept.ID),
		Description:  fmt.Sprintf("Created enterprise department %s for enterprise %d", req.Name, req.EnterpriseID),
		Status:       "success",
	})

	return createdDept, nil
}

// GetEnterpriseDepartmentByID gets an enterprise department by ID
func (s *EnterpriseService) GetEnterpriseDepartmentByID(ctx context.Context, id int) (*models.EnterpriseDepartment, error) {
	dept, err := s.enterpriseRepo.GetDepartmentByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("enterprise department not found: %w", err)
	}

	return dept, nil
}

// UpdateEnterpriseDepartment updates an enterprise department
func (s *EnterpriseService) UpdateEnterpriseDepartment(ctx context.Context, id int, req *models.EnterpriseDepartmentRequest, operatorID int) (*models.EnterpriseDepartment, error) {
	// Get existing department
	dept, err := s.enterpriseRepo.GetDepartmentByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("enterprise department not found: %w", err)
	}

	// Validate parent department if specified
	if req.ParentID != nil {
		parent, err := s.enterpriseRepo.GetDepartmentByID(ctx, *req.ParentID)
		if err != nil {
			return nil, fmt.Errorf("parent department not found: %w", err)
		}
		if parent.EnterpriseID != dept.EnterpriseID {
			return nil, fmt.Errorf("parent department belongs to different enterprise")
		}
		// Prevent circular reference
		if *req.ParentID == id {
			return nil, fmt.Errorf("department cannot be its own parent")
		}
	}

	// Update fields
	dept.Name = req.Name
	dept.ParentID = req.ParentID
	dept.SortOrder = req.SortOrder
	dept.ManagerID = req.ManagerID
	dept.Description = req.Description
	dept.EmployeeCount = req.EmployeeCount
	dept.Status = req.Status
	dept.UpdatedBy = &operatorID

	// Update department
	updatedDept, err := s.enterpriseRepo.UpdateDepartment(ctx, dept)
	if err != nil {
		return nil, fmt.Errorf("failed to update enterprise department: %w", err)
	}

	// Log the update event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "update",
		ResourceType: "enterprise_department",
		ResourceID:   fmt.Sprintf("%d", id),
		Description:  fmt.Sprintf("Updated enterprise department %s", dept.Name),
		Status:       "success",
	})

	return updatedDept, nil
}

// DeleteEnterpriseDepartment soft deletes an enterprise department
func (s *EnterpriseService) DeleteEnterpriseDepartment(ctx context.Context, id int, operatorID int) error {
	// Get department for logging
	dept, err := s.enterpriseRepo.GetDepartmentByID(ctx, id)
	if err != nil {
		return fmt.Errorf("enterprise department not found: %w", err)
	}

	// Check if department has children (you might want to prevent deletion)
	children, err := s.GetEnterpriseDepartments(ctx, dept.EnterpriseID)
	if err != nil {
		return fmt.Errorf("failed to check department children: %w", err)
	}

	for _, child := range children {
		if child.ParentID != nil && *child.ParentID == id {
			return fmt.Errorf("cannot delete department with child departments")
		}
	}

	// Delete department
	err = s.enterpriseRepo.DeleteDepartment(ctx, id)
	if err != nil {
		return fmt.Errorf("failed to delete enterprise department: %w", err)
	}

	// Log the deletion event
	s.auditLogger.LogEvent(&models.AuditEventData{
		UserID:       &operatorID,
		Action:       "delete",
		ResourceType: "enterprise_department",
		ResourceID:   fmt.Sprintf("%d", id),
		Description:  fmt.Sprintf("Deleted enterprise department %s", dept.Name),
		Status:       "success",
	})

	return nil
}

// GetEnterpriseDepartments lists departments for an enterprise
func (s *EnterpriseService) GetEnterpriseDepartments(ctx context.Context, enterpriseID int) ([]*models.EnterpriseDepartment, error) {
	// Validate enterprise exists
	_, err := s.enterpriseRepo.GetByID(ctx, enterpriseID)
	if err != nil {
		return nil, fmt.Errorf("enterprise not found: %w", err)
	}

	departments, err := s.enterpriseRepo.GetDepartments(ctx, enterpriseID)
	if err != nil {
		return nil, fmt.Errorf("failed to get enterprise departments: %w", err)
	}

	return departments, nil
}

// GetEnterpriseDepartmentStats returns department statistics for an enterprise
func (s *EnterpriseService) GetEnterpriseDepartmentStats(ctx context.Context, enterpriseID int) (*models.EnterpriseDepartmentStats, error) {
	// Validate enterprise exists
	_, err := s.enterpriseRepo.GetByID(ctx, enterpriseID)
	if err != nil {
		return nil, fmt.Errorf("enterprise not found: %w", err)
	}

	stats, err := s.enterpriseRepo.GetDepartmentStats(ctx, enterpriseID)
	if err != nil {
		return nil, fmt.Errorf("failed to get enterprise department statistics: %w", err)
	}

	return stats, nil
}

// Helper methods

// generateRandomPassword generates a secure random password
func (s *EnterpriseService) generateRandomPassword() (string, error) {
	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*"
	const length = 12

	password := make([]byte, length)
	for i := range password {
		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(charset))))
		if err != nil {
			return "", err
		}
		password[i] = charset[num.Int64()]
	}

	return string(password), nil
}

// SendWelcomeEmail sends welcome email to new enterprise user
func (s *EnterpriseService) SendWelcomeEmail(user *models.EnterpriseUser, password string) error {
	// Email service implementation would go here
	// For now, this is a placeholder
	return nil
}

// IsAccountExpired checks if an enterprise user account is expired (placeholder for future implementation)
func (s *EnterpriseService) IsAccountExpired(user *models.EnterpriseUser) bool {
	// Enterprise users don't have expiration by default in this implementation
	// This can be extended in the future
	return false
}

File: frontend/src/services/enterpriseService.ts
import api from './api';

// Enterprise data types based on backend models
export interface Enterprise {
  id: number;
  name: string;
  code: string;
  description?: string;
  industry_type?: string;
  industry_type_text?: string;
  business_type: 'individual' | 'partnership' | 'corporation' | 'llc';
  business_type_text: string;
  registration_number?: string;
  tax_id?: string;
  legal_representative?: string;
  contact_email?: string;
  contact_phone?: string;
  address?: string;
  city?: string;
  province?: string;
  postal_code?: string;
  website?: string;
  status: 'active' | 'inactive' | 'suspended';
  status_text: string;
  user_count?: number;
  department_count?: number;
  created_by?: number;
  updated_by?: number;
  created_at: string;
  updated_at: string;
}

// Enterprise User data types
export interface EnterpriseUser {
  id: number;
  enterprise_id: number;
  username: string;
  email?: string;
  name: string;
  phone?: string;
  position?: string;
  department_id?: number;
  department_name?: string;
  is_primary_contact: boolean;
  access_level: number;
  status: 'active' | 'inactive' | 'suspended';
  status_text: string;
  last_login_at?: string;
  created_at: string;
  updated_at: string;
}

// Enterprise Department data types
export interface EnterpriseDepartment {
  id: number;
  enterprise_id: number;
  name: string;
  code?: string;
  description?: string;
  parent_id?: number;
  parent_name?: string;
  level: number;
  path?: string;
  status: 'active' | 'inactive';
  status_text: string;
  user_count: number;
  created_at: string;
  updated_at: string;
}

// Request types
export interface EnterpriseRequest {
  name: string;
  code: string;
  description?: string;
  industry_type?: string;
  business_type: 'individual' | 'partnership' | 'corporation' | 'llc';
  registration_number?: string;
  tax_id?: string;
  legal_representative?: string;
  contact_email?: string;
  contact_phone?: string;
  address?: string;
  city?: string;
  province?: string;
  postal_code?: string;
  website?: string;
  status: 'active' | 'inactive' | 'suspended';
}

export interface EnterpriseUpdateRequest {
  name?: string;
  code?: string;
  description?: string;
  industry_type?: string;
  business_type?: 'individual' | 'partnership' | 'corporation' | 'llc';
  registration_number?: string;
  tax_id?: string;
  legal_representative?: string;
  contact_email?: string;
  contact_phone?: string;
  address?: string;
  city?: string;
  province?: string;
  postal_code?: string;
  website?: string;
  status?: 'active' | 'inactive' | 'suspended';
}

export interface EnterpriseUserRequest {
  username: string;
  name: string;
  email?: string;
  phone?: string;
  position?: string;
  department_id?: number;
  is_primary_contact?: boolean;
  access_level?: number;
  status?: 'active' | 'inactive' | 'suspended';
}

export interface EnterpriseUserUpdateRequest {
  enterprise_id?: number;
  username: string;
  name: string;
  email: string;
  phone?: string;
  position?: string;
  department_id?: number;
  is_primary_contact?: boolean;
  access_level: number;
  status: 'active' | 'inactive' | 'archived';
}

export interface EnterpriseDepartmentRequest {
  name: string;
  code?: string;
  description?: string;
  parent_id?: number;
  status?: 'active' | 'inactive';
}

// API Response types
export interface PaginatedResponse<T> {
  data: T[];
  pagination: {
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
    has_next: boolean;
    has_prev: boolean;
  };
}

export interface EnterpriseStats {
  total_enterprises: number;
  active_enterprises: number;
  inactive_enterprises: number;
  suspended_enterprises: number;
  total_users: number;
  total_departments: number;
  enterprises_by_business_type: Array<{
    business_type: string;
    count: number;
  }>;
  enterprises_by_industry: Array<{
    industry_type: string;
    count: number;
  }>;
}

class EnterpriseService {
  private readonly API_BASE_URL = '/enterprises';

  // 统一的 API 响应处理函数
  private handleApiResponse<T>(response: any): T {
    
    if (!response) {
      return {
        data: [],
        pagination: {
          page: 1,
          page_size: 20,
          total: 0,
          total_pages: 0,
          has_next: false,
          has_prev: false
        }
      } as T;
    }

    // 新增：如果响应本身就是数组（拦截器已解包为数组），包装为统一的分页结构
    if (Array.isArray(response)) {
      return {
        data: response,
        pagination: {
          page: 1,
          page_size: response.length,
          total: response.length,
          total_pages: 1,
          has_next: false,
          has_prev: false
        }
      } as T;
    }

    // 处理嵌套的 data 响应格式（拦截器已解包response.data，所以原来的data.data变成了data）
    if (response && typeof response === 'object' && Array.isArray(response.data)) {
      return {
        data: response.data,
        pagination: response.pagination
      } as T;
    }

    // 处理API返回的 { data: [] } 直接数组格式（拦截器已解包）
    // 这种情况在上面的条件中已处理，这里保留作为后备

    // 直接返回响应数据，API拦截器已经处理了标准响应格式
    return response as T;
  }

  // 获取企业列表
  async getEnterprises(page: number = 1, pageSize: number = 20, filters?: any): Promise<PaginatedResponse<Enterprise>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        page_size: pageSize.toString(),
        ...filters
      });
      const response = await api.get(`${this.API_BASE_URL}?${params}`);
      const result = this.handleApiResponse<PaginatedResponse<Enterprise>>(response);
      return result;
    } catch (error) {
      console.error('❌ 获取企业列表失败:', error);
      throw error;
    }
  }

  // 获取单个企业详情
  async getEnterprise(id: number): Promise<Enterprise> {
    try {
      const response = await api.get(`${this.API_BASE_URL}/${id}`);
      const result = this.handleApiResponse<Enterprise>(response);
      return result;
    } catch (error) {
      console.error('❌ 获取企业详情失败:', error);
      throw error;
    }
  }

  // 创建企业
  async createEnterprise(enterprise: EnterpriseRequest): Promise<Enterprise> {
    try {
      const response = await api.post(this.API_BASE_URL, enterprise);
      const result = this.handleApiResponse<Enterprise>(response);
      return result;
    } catch (error) {
      console.error('❌ 创建企业失败:', error);
      throw error;
    }
  }

  // 更新企业
  async updateEnterprise(id: number, enterprise: EnterpriseUpdateRequest): Promise<Enterprise> {
    try {
      const response = await api.put(`${this.API_BASE_URL}/${id}`, enterprise);
      const result = this.handleApiResponse<Enterprise>(response);
      return result;
    } catch (error) {
      console.error('❌ 更新企业失败:', error);
      throw error;
    }
  }

  // 删除企业
  async deleteEnterprise(id: number): Promise<void> {
    try {
      await api.delete(`${this.API_BASE_URL}/${id}`);
    } catch (error) {
      console.error('❌ 删除企业失败:', error);
      throw error;
    }
  }

  // 获取企业统计信息
  async getEnterpriseStats(): Promise<EnterpriseStats> {
    try {
      const response = await api.get(`${this.API_BASE_URL}/stats`);
      const result = this.handleApiResponse<EnterpriseStats>(response);
      return result;
    } catch (error) {
      console.error('❌ 获取企业统计信息失败:', error);
      // 返回默认统计信息
      return {
        total_enterprises: 0,
        active_enterprises: 0,
        inactive_enterprises: 0,
        suspended_enterprises: 0,
        total_users: 0,
        total_departments: 0,
        enterprises_by_business_type: [],
        enterprises_by_industry: []
      };
    }
  }

  // 获取企业用户列表
  async getEnterpriseUsers(enterpriseId: number, page: number = 1, pageSize: number = 20, filters?: any): Promise<PaginatedResponse<EnterpriseUser>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        page_size: pageSize.toString(),
        ...filters
      });
      const response = await api.get(`${this.API_BASE_URL}/${enterpriseId}/users?${params}`);
      const result = this.handleApiResponse<PaginatedResponse<EnterpriseUser>>(response);
      return result;
    } catch (error) {
      console.error('❌ 获取企业用户列表失败:', error);
      throw error;
    }
  }

  // 获取单个企业用户详情
  async getEnterpriseUser(enterpriseId: number, userId: number): Promise<EnterpriseUser> {
    try {
      const response = await api.get(`${this.API_BASE_URL}/${enterpriseId}/users/${userId}`);
      const result = this.handleApiResponse<EnterpriseUser>(response);
      return result;
    } catch (error) {
      console.error('❌ 获取企业用户详情失败:', error);
      throw error;
    }
  }

  // 创建企业用户
  async createEnterpriseUser(enterpriseId: number, user: EnterpriseUserRequest): Promise<{
    user: EnterpriseUser;
    generated_password?: string;
  }> {
    try {
      const response = await api.post(`${this.API_BASE_URL}/${enterpriseId}/users`, user);
      const result = this.handleApiResponse<{
        user: EnterpriseUser;
        generated_password?: string;
      }>(response);
      return result;
    } catch (error) {
      console.error('❌ 创建企业用户失败:', error);
      throw error;
    }
  }

  // 更新企业用户
  async updateEnterpriseUser(enterpriseId: number, userId: number, user: Partial<EnterpriseUserRequest>): Promise<EnterpriseUser> {
    try {
      const response = await api.put(`${this.API_BASE_URL}/${enterpriseId}/users/${userId}`, user);
      const result = this.handleApiResponse<EnterpriseUser>(response);
      return result;
    } catch (error) {
      console.error('❌ 更新企业用户失败:', error);
      throw error;
    }
  }

  // 获取企业部门列表
  async getEnterpriseDepartments(enterpriseId: number, page: number = 1, pageSize: number = 20, filters?: any): Promise<PaginatedResponse<EnterpriseDepartment>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        page_size: pageSize.toString(),
        ...filters
      });
      const response = await api.get(`${this.API_BASE_URL}/${enterpriseId}/departments?${params}`);
      const result = this.handleApiResponse<PaginatedResponse<EnterpriseDepartment>>(response);
      return result;
    } catch (error) {
      console.error('❌ 获取企业部门列表失败:', error);
      throw error;
    }
  }

  // 创建企业部门
  async createEnterpriseDepartment(enterpriseId: number, department: EnterpriseDepartmentRequest): Promise<EnterpriseDepartment> {
    try {
      const response = await api.post(`${this.API_BASE_URL}/${enterpriseId}/departments`, department);
      const result = this.handleApiResponse<EnterpriseDepartment>(response);
      return result;
    } catch (error) {
      console.error('❌ 创建企业部门失败:', error);
      throw error;
    }
  }

  // 更新企业部门
  async updateEnterpriseDepartment(enterpriseId: number, departmentId: number, department: EnterpriseDepartmentRequest): Promise<EnterpriseDepartment> {
    try {
      const response = await api.put(`${this.API_BASE_URL}/${enterpriseId}/departments/${departmentId}`, department);
      const result = this.handleApiResponse<EnterpriseDepartment>(response);
      return result;
    } catch (error) {
      console.error('❌ 更新企业部门失败:', error);
      throw error;
    }
  }

  // 删除企业部门
  async deleteEnterpriseDepartment(enterpriseId: number, departmentId: number): Promise<void> {
    try {
      await api.delete(`${this.API_BASE_URL}/${enterpriseId}/departments/${departmentId}`);
    } catch (error) {
      console.error('❌ 删除企业部门失败:', error);
      throw error;
    }
  }

  // 获取业务类型选项
  getBusinessTypeOptions(): Array<{ value: string; label: string }> {
    return [
      { value: 'individual', label: '个人' },
      { value: 'partnership', label: '合伙制' },
      { value: 'corporation', label: '股份制' },
      { value: 'llc', label: '有限责任公司' }
    ];
  }

  // 获取状态选项
  getStatusOptions(): Array<{ value: string; label: string }> {
    return [
      { value: 'active', label: '活跃' },
      { value: 'inactive', label: '非活跃' },
      { value: 'suspended', label: '暂停' }
    ];
  }

  // 获取访问级别选项
  getAccessLevelOptions(): Array<{ value: number; label: string }> {
    return [
      { value: 1, label: '基础权限' },
      { value: 2, label: '一般权限' },
      { value: 3, label: '中级权限' },
      { value: 4, label: '高级权限' },
      { value: 5, label: '最高权限' }
    ];
  }
}

export default new EnterpriseService();